### PR TITLE
LibDiff+diff: Coalesce adjacent changes into a single Hunk, show line ranges

### PR DESCRIPTION
**Pull request by mustafaquraish**
After a little refactor reversing the order in which we ran the diff algorithm, now we can neatly coalesce changes into hunks. The `diff` utility now also shows the correct line range(s) for the hunks.

### CODE REVIEW

File: Userland/Libraries/LibDiff/Generator.cpp [hunk 1]
```diff
@@ -63,21 +63,43 @@ Vector<Hunk> from_text(StringView const& old_text, StringView const& new_text)
...
+            if (in_hunk) {
+                if (cur_hunk.added_lines.size() > 0)
+                    cur_hunk.target_start_line++;
+                if (cur_hunk.removed_lines.size() > 0)
+                    cur_hunk.original_start_line++;
+                hunks.append(cur_hunk);
+                in_hunk = false;
+            }
         }
     }
```
Comment (BenWiederhake) [inline Userland/Libraries/LibDiff/Generator.cpp:105]: Off-by-one error: If `in_hunk = true` in the end, then that hunk is lost.
Suggestion: Create a `flush_hunk` that contains the entire `if (in_hunk) { … }` logic, and call it both in the `else` branch, as well as after the `while` loop.

Also, to make sure that this type of error never crops up again: How do you feel about writing a few tests? I strongly recommending *first* writing a test case that shows that the current code is broken, and only *then* fixing the code. Of course, the commits should not reflect that.

File: Userland/Utilities/diff.cpp [hunk 2]
```diff
@@ -37,11 +37,29 @@ int main(int argc, char** argv)
...
+        auto num_removed = hunk.removed_lines.size();
+
+        StringBuilder sb;
+        // Source line(s)
+        sb.appendff("{}", original_start);
+        if (num_removed > 1)
+            sb.appendff(",{}", original_start + num_removed - 1);
+
+        // Action
+        sb.append(num_added > 0 && num_removed > 0 ? "c" : num_added > 0 ? "a"
```
Comment (BenWiederhake) [inline Userland/Utilities/diff.cpp:52]: Nested ternaries are evil! :P
Please use `if`s here, especially if it doesn't comfortably fit in a line anyway.

File: Userland/Libraries/LibDiff/Generator.cpp [hunk 3]
```diff
@@ -41,47 +41,46 @@ Vector<Hunk> from_text(StringView const& old_text, StringView const& new_text)
...
-            --i;
-            --j;
+            ++i;
+        } else {
+            ++i;
+            ++j;
         }
     }
 
-    hunks.reverse();
```
Comment (BenWiederhake) [inline Userland/Libraries/LibDiff/Generator.cpp:84]: Sneaky: This removes the last user of `Vector::reverse`.

File: Userland/Utilities/diff.cpp [hunk 4]
```diff
@@ -37,11 +37,29 @@ int main(int argc, char** argv)
 
     auto hunks = Diff::from_text(file1->read_all(), file2->read_all());
     for (const auto& hunk : hunks) {
```
Comment (BenWiederhake) [inline Userland/Utilities/diff.cpp:39]: Nitpick about the commit message: I personally prefer just `diff: Blah blah` over `Utilities/diff: Blah blah`, but I guess this is fine. Feel free to click "resolve" without changing it, if you prefer it the current way.

**Review by BenWiederhake — CHANGES_REQUESTED** [review]
A small off-by-one error, and a tiny style thing. Apart from that: Yup, it's nice!

File: Userland/Libraries/LibDiff/Generator.cpp (hunk 3 above)
Comment (mustafaquraish) [reply Userland/Libraries/LibDiff/Generator.cpp:84]: Indeed, this is sneaky, but despite the dead-code policy I think that `Vector::reverse()` is a sensible API to have (and it's like 2 lines anyway). If you think I should remove it again, I'll do it ¯\_(ツ)_/¯

File: Userland/Libraries/LibDiff/Generator.cpp (hunk 1 above)
Comment (mustafaquraish) [reply Userland/Libraries/LibDiff/Generator.cpp]: Adding tests is definitely a good idea, though I don't have much time these days to sit down and write a bunch of them. I'll leave for a future PR when I get some free time!

**Comment by mustafaquraish** [comment]
cc @bgianfo since you merged the last one, and have any thoughts here

**Pull request merged**